### PR TITLE
Bump needed `{withr}` version

### LINKED ISSUE #1404: Bump withr dependency to 2.5.0 and use local_tempfile(lines = ) in tests
This is a convenient improvement to `local_tempfile()` that's available since withr 2.5.0 on CRAN since March. By the time we release again, it should be mature enough that we are happy to depend on it.

_Originally posted by @MichaelChirico in https://github.com/r-lib/lintr/pull/1403#discussion_r900374747_

### PR DESCRIPTION
**Pull request by IndrajeetPatil**
In order to use the `lines` parameter from `withr::local_tempfile()`.

Closes #1404

### CODE REVIEW

File: tests/testthat/test-rstudio_markers.R [hunk 1]
```diff
@@ -100,11 +100,9 @@ test_that("it returns an empty list of markers if there are no lints", {
...
 test_that("rstudio_source_markers apply to print within rstudio", {
   withr::local_options(lintr.rstudio_source_markers = TRUE)
-  # TODO(@michaelchirico): Recent (as of this writing) withr v2.5.0 supports local_tempfile(lines = l) to simplify this
-  tmp <- withr::local_tempfile()
-  writeLines("1:ncol(x)", tmp)
-  empty <- withr::local_tempfile()
-  file.create(empty)
+
+  tmp <- withr::local_tempfile(lines = "1:ncol(x)")
+  empty <- withr::local_tempfile(lines = "")
```
Comment (AshesITR) [inline tests/testthat/test-rstudio_markers.R:105]: Wouldn't that be a file containing a single newline character instead of an empty file?

Comment (IndrajeetPatil) [reply tests/testthat/test-rstudio_markers.R]: Good catch.

Fixed it:

``` r
empty <- withr::local_tempfile()
file.create(empty)
#> [1] TRUE
readLines(empty)
#> character(0)

readLines(withr::local_tempfile(lines = character(0L)))
#> character(0)
```

<sup>Created on 2022-06-18 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

File: tests/testthat/test-get_source_expressions.R [hunk 2]
```diff
@@ -292,8 +290,7 @@ skip_if_not_installed("patrick")
 #   fail on files where the XML content is xml_missing;
 #   the main linter test files provide more thorough
 #   evidence that things are working as intended.
-bad_source <- withr::local_tempfile()
-writeLines("a <- 1L\nb <- 2L", bad_source)
+bad_source <- withr::local_tempfile(lines = "a <- 1L\nb <- 2L")
```
Comment (AshesITR) [inline tests/testthat/test-get_source_expressions.R:293]: This will add a terminal newline, IINM.

**Review by AshesITR — CHANGES_REQUESTED** [review]
Please also refactor the usages at

 - test-parse_exclusions.R:153
 - test-get_source_expressions.R:266
 - test-get_source_expressions.R:282
 - test-get_source_expressions.R:294

File: tests/testthat/test-get_source_expressions.R (hunk 2 above)
Comment (IndrajeetPatil) [reply tests/testthat/test-get_source_expressions.R:293]: Are you sure about that? Am I missing something here?

``` r
bad_source1 <- withr::local_tempfile()
writeLines("a <- 1L\nb <- 2L", bad_source1)

bad_source2 <- withr::local_tempfile(lines = "a <- 1L\nb <- 2L")

waldo::compare(readLines(bad_source1), readLines(bad_source2))
#> v No differences
```

<sup>Created on 2022-06-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

Comment (AshesITR) [reply tests/testthat/test-get_source_expressions.R:292]: Ah, sorry, `writeLines()` adds the final newline implicitly.
`cat()` wouldn't have.
But then we could use `lines = c("a <- 1L", "b <- 2L")` for a little improved readability.

Comment (IndrajeetPatil) [reply tests/testthat/test-get_source_expressions.R]: Okay, done.

**Comment by IndrajeetPatil** [comment]
> Please also refactor the usages at
> 
> * test-parse_exclusions.R:153
> * test-get_source_expressions.R:266
> * test-get_source_expressions.R:282
> * test-get_source_expressions.R:294

Done.

**Review by AshesITR — APPROVED** [review]
Thanks!

**Comment by MichaelChirico** [comment]
Thanks @IndrajeetPatil !!

@jimhester could you add Indrajeet as a collaborator?

**Comment by IndrajeetPatil** [comment]
Thanks, @MichaelChirico and @jimhester! 😊

**Pull request merged**